service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java

```diff
@@ -55,68 +52,91 @@ public PaymentHandler(String payoneInterfaceName, String tenantName,
-                    return new PaymentHandleResult(HttpStatusCode.OK_200);
-                } catch (final ConcurrentModificationException cme) {
-                    logger.warn(tenantNameKeyValue, "Exception on dispatchPayment with id [{}]. Retry {} of {}", paymentId, i + 1, RETRIES_LIMIT);
-                    Thread.sleep(RETRY_DELAY);
-                }
-            }
-
-            logger.error(
-                tenantNameKeyValue,
-                "The payment [{}] couldn't be processed after {} retries", paymentId, RETRIES_LIMIT);
-            return new PaymentHandleResult(HttpStatusCode.ACCEPTED_202,
-                    format("The payment couldn't be processed after %s retries", RETRIES_LIMIT));
-
+            return attemptToProcessPayment(paymentId, 0);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            return handleConcurrentModificationException(paymentId, concurrentModificationException);
         } catch (final NotFoundException | NoCartLikeFoundException e) {
             return handleNotFoundException(paymentId, e);
         } catch (final ErrorResponseException e) {
             return errorResponseHandler(e, paymentId);
-        } catch (final CompletionException e) {
-            return completionExceptionHandler(e, paymentId);
         } catch (final Exception e) {
-            return handleThrowableInResponse(e, paymentId);
+            return handleException(e, paymentId);
+        }
+    }
+
+    private PaymentHandleResult attemptToProcessPayment(
+        @Nonnull final String paymentId,
+        final int currentAttemptCount) throws InterruptedException {
+
+        try {
+            return processPayment(paymentId);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            if (currentAttemptCount == RETRIES_LIMIT - 1) {
+                throw concurrentModificationException;
+            } else {
+                Thread.sleep(RETRY_DELAY);
+                return attemptToProcessPayment(paymentId, currentAttemptCount + 1);
+            }
         }
     }
 
-     private PaymentHandleResult handleNotFoundException(@Nonnull String paymentId, @Nonnull Exception exception ) {
+    private PaymentHandleResult processPayment(@Nonnull final String paymentId) {
+        final PaymentWithCartLike paymentWithCartLike =
+            commercetoolsQueryExecutor
+                .getPaymentWithCartLike(paymentId);
+
+        final String paymentInterface = paymentWithCartLike
+            .getPayment()
+            .getPaymentMethodInfo()
+            .getPaymentInterface();
+
+        if (!payoneInterfaceName.equals(paymentInterface)) {
+            logger.warn(tenantNameKeyValue,
+                "Wrong payment interface name: expected [{}], found [{}]",
+                payoneInterfaceName, paymentInterface);
+            return new PaymentHandleResult(HttpStatusCode.BAD_REQUEST_400);
```

Comment:
Should you also inform the customer about the error?

Reply:
I removed the warning here at all and put the error body in the response: 0e934f2